{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import contextily as ctxl\n",
    "import xyzservices.providers as xyz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "%matplotlib inline\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_context('notebook')\n",
    "sns.set_style('ticks', \n",
    "    {\n",
    "        'axes.grid': True, \n",
    "        'axes.linewidth': '1',\n",
    "        'grid.color': '0.5',\n",
    "        'grid.linestyle': u':',\n",
    "        'legend.frameon': True,\n",
    "    })\n",
    "xkcd = sns.xkcd_rgb\n",
    "\n",
    "rc_opt = {\n",
    "    \"font.family\": \"serif\",\n",
    "    \"font.size\": 15,\n",
    "    \"text.usetex\": True,\n",
    "    'text.latex.preamble': r\"\\usepackage{libertine} \\usepackage[libertine]{newtxmath}\",\n",
    "}\n",
    "mpl.rcParams.update(rc_opt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/loh/.local/conda/envs/gis/lib/python3.10/site-packages/contextily/tile.py:581: UserWarning: The inferred zoom level of 27 is not valid for the current tile provider (valid zooms: 0 - 19).\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "ename": "SSLError",
     "evalue": "HTTPSConnectionPool(host='a.tile.openstreetmap.org', port=443): Max retries exceeded with url: /19/262144/262143.png (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:997)')))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mSSLCertVerificationError\u001b[0m                  Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/urllib3/connectionpool.py:703\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    704\u001b[0m     conn,\n\u001b[1;32m    705\u001b[0m     method,\n\u001b[1;32m    706\u001b[0m     url,\n\u001b[1;32m    707\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    708\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    709\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    710\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    711\u001b[0m )\n\u001b[1;32m    713\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    714\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    715\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    716\u001b[0m \u001b[39m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/urllib3/connectionpool.py:386\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    385\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 386\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_validate_conn(conn)\n\u001b[1;32m    387\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    388\u001b[0m     \u001b[39m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/urllib3/connectionpool.py:1042\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1041\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mgetattr\u001b[39m(conn, \u001b[39m\"\u001b[39m\u001b[39msock\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):  \u001b[39m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m-> 1042\u001b[0m     conn\u001b[39m.\u001b[39;49mconnect()\n\u001b[1;32m   1044\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m conn\u001b[39m.\u001b[39mis_verified:\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/urllib3/connection.py:414\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    412\u001b[0m     context\u001b[39m.\u001b[39mload_default_certs()\n\u001b[0;32m--> 414\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msock \u001b[39m=\u001b[39m ssl_wrap_socket(\n\u001b[1;32m    415\u001b[0m     sock\u001b[39m=\u001b[39;49mconn,\n\u001b[1;32m    416\u001b[0m     keyfile\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mkey_file,\n\u001b[1;32m    417\u001b[0m     certfile\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcert_file,\n\u001b[1;32m    418\u001b[0m     key_password\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mkey_password,\n\u001b[1;32m    419\u001b[0m     ca_certs\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_certs,\n\u001b[1;32m    420\u001b[0m     ca_cert_dir\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_cert_dir,\n\u001b[1;32m    421\u001b[0m     ca_cert_data\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mca_cert_data,\n\u001b[1;32m    422\u001b[0m     server_hostname\u001b[39m=\u001b[39;49mserver_hostname,\n\u001b[1;32m    423\u001b[0m     ssl_context\u001b[39m=\u001b[39;49mcontext,\n\u001b[1;32m    424\u001b[0m     tls_in_tls\u001b[39m=\u001b[39;49mtls_in_tls,\n\u001b[1;32m    425\u001b[0m )\n\u001b[1;32m    427\u001b[0m \u001b[39m# If we're using all defaults and the connection\u001b[39;00m\n\u001b[1;32m    428\u001b[0m \u001b[39m# is TLSv1 or TLSv1.1 we throw a DeprecationWarning\u001b[39;00m\n\u001b[1;32m    429\u001b[0m \u001b[39m# for the host.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/urllib3/util/ssl_.py:449\u001b[0m, in \u001b[0;36mssl_wrap_socket\u001b[0;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password, ca_cert_data, tls_in_tls)\u001b[0m\n\u001b[1;32m    448\u001b[0m \u001b[39mif\u001b[39;00m send_sni:\n\u001b[0;32m--> 449\u001b[0m     ssl_sock \u001b[39m=\u001b[39m _ssl_wrap_socket_impl(\n\u001b[1;32m    450\u001b[0m         sock, context, tls_in_tls, server_hostname\u001b[39m=\u001b[39;49mserver_hostname\n\u001b[1;32m    451\u001b[0m     )\n\u001b[1;32m    452\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/urllib3/util/ssl_.py:493\u001b[0m, in \u001b[0;36m_ssl_wrap_socket_impl\u001b[0;34m(sock, ssl_context, tls_in_tls, server_hostname)\u001b[0m\n\u001b[1;32m    492\u001b[0m \u001b[39mif\u001b[39;00m server_hostname:\n\u001b[0;32m--> 493\u001b[0m     \u001b[39mreturn\u001b[39;00m ssl_context\u001b[39m.\u001b[39;49mwrap_socket(sock, server_hostname\u001b[39m=\u001b[39;49mserver_hostname)\n\u001b[1;32m    494\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/ssl.py:513\u001b[0m, in \u001b[0;36mSSLContext.wrap_socket\u001b[0;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)\u001b[0m\n\u001b[1;32m    507\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mwrap_socket\u001b[39m(\u001b[39mself\u001b[39m, sock, server_side\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m    508\u001b[0m                 do_handshake_on_connect\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m,\n\u001b[1;32m    509\u001b[0m                 suppress_ragged_eofs\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m,\n\u001b[1;32m    510\u001b[0m                 server_hostname\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, session\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[1;32m    511\u001b[0m     \u001b[39m# SSLSocket class handles server_hostname encoding before it calls\u001b[39;00m\n\u001b[1;32m    512\u001b[0m     \u001b[39m# ctx._wrap_socket()\u001b[39;00m\n\u001b[0;32m--> 513\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msslsocket_class\u001b[39m.\u001b[39;49m_create(\n\u001b[1;32m    514\u001b[0m         sock\u001b[39m=\u001b[39;49msock,\n\u001b[1;32m    515\u001b[0m         server_side\u001b[39m=\u001b[39;49mserver_side,\n\u001b[1;32m    516\u001b[0m         do_handshake_on_connect\u001b[39m=\u001b[39;49mdo_handshake_on_connect,\n\u001b[1;32m    517\u001b[0m         suppress_ragged_eofs\u001b[39m=\u001b[39;49msuppress_ragged_eofs,\n\u001b[1;32m    518\u001b[0m         server_hostname\u001b[39m=\u001b[39;49mserver_hostname,\n\u001b[1;32m    519\u001b[0m         context\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m,\n\u001b[1;32m    520\u001b[0m         session\u001b[39m=\u001b[39;49msession\n\u001b[1;32m    521\u001b[0m     )\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/ssl.py:1071\u001b[0m, in \u001b[0;36mSSLSocket._create\u001b[0;34m(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)\u001b[0m\n\u001b[1;32m   1070\u001b[0m             \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mdo_handshake_on_connect should not be specified for non-blocking sockets\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m-> 1071\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdo_handshake()\n\u001b[1;32m   1072\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mOSError\u001b[39;00m, \u001b[39mValueError\u001b[39;00m):\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/ssl.py:1342\u001b[0m, in \u001b[0;36mSSLSocket.do_handshake\u001b[0;34m(self, block)\u001b[0m\n\u001b[1;32m   1341\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msettimeout(\u001b[39mNone\u001b[39;00m)\n\u001b[0;32m-> 1342\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sslobj\u001b[39m.\u001b[39;49mdo_handshake()\n\u001b[1;32m   1343\u001b[0m \u001b[39mfinally\u001b[39;00m:\n",
      "\u001b[0;31mSSLCertVerificationError\u001b[0m: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:997)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/requests/adapters.py:489\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 489\u001b[0m     resp \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    490\u001b[0m         method\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mmethod,\n\u001b[1;32m    491\u001b[0m         url\u001b[39m=\u001b[39;49murl,\n\u001b[1;32m    492\u001b[0m         body\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mbody,\n\u001b[1;32m    493\u001b[0m         headers\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    494\u001b[0m         redirect\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    495\u001b[0m         assert_same_host\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    496\u001b[0m         preload_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    497\u001b[0m         decode_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    498\u001b[0m         retries\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmax_retries,\n\u001b[1;32m    499\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[1;32m    500\u001b[0m     )\n\u001b[1;32m    502\u001b[0m \u001b[39m# Send the request.\u001b[39;00m\n\u001b[1;32m    503\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/urllib3/connectionpool.py:787\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    785\u001b[0m     e \u001b[39m=\u001b[39m ProtocolError(\u001b[39m\"\u001b[39m\u001b[39mConnection aborted.\u001b[39m\u001b[39m\"\u001b[39m, e)\n\u001b[0;32m--> 787\u001b[0m retries \u001b[39m=\u001b[39m retries\u001b[39m.\u001b[39;49mincrement(\n\u001b[1;32m    788\u001b[0m     method, url, error\u001b[39m=\u001b[39;49me, _pool\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m, _stacktrace\u001b[39m=\u001b[39;49msys\u001b[39m.\u001b[39;49mexc_info()[\u001b[39m2\u001b[39;49m]\n\u001b[1;32m    789\u001b[0m )\n\u001b[1;32m    790\u001b[0m retries\u001b[39m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/urllib3/util/retry.py:592\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    591\u001b[0m \u001b[39mif\u001b[39;00m new_retry\u001b[39m.\u001b[39mis_exhausted():\n\u001b[0;32m--> 592\u001b[0m     \u001b[39mraise\u001b[39;00m MaxRetryError(_pool, url, error \u001b[39mor\u001b[39;00m ResponseError(cause))\n\u001b[1;32m    594\u001b[0m log\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mIncremented Retry for (url=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m): \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m, url, new_retry)\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPSConnectionPool(host='a.tile.openstreetmap.org', port=443): Max retries exceeded with url: /19/262144/262143.png (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:997)')))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mSSLError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m/home/loh/Workspace/methane_flux/notebook/retrieve_map.ipynb Cell 3\u001b[0m in \u001b[0;36m<cell line: 6>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/loh/Workspace/methane_flux/notebook/retrieve_map.ipynb#X13sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m ax \u001b[39m=\u001b[39m fig\u001b[39m.\u001b[39madd_subplot(\u001b[39m111\u001b[39m)\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/loh/Workspace/methane_flux/notebook/retrieve_map.ipynb#X13sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m extent \u001b[39m=\u001b[39m (\u001b[39m-\u001b[39m\u001b[39m12600000\u001b[39m, \u001b[39m-\u001b[39m\u001b[39m10300000\u001b[39m, \u001b[39m1800000\u001b[39m, \u001b[39m3800000\u001b[39m)\n\u001b[0;32m----> <a href='vscode-notebook-cell:/home/loh/Workspace/methane_flux/notebook/retrieve_map.ipynb#X13sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m ctxl\u001b[39m.\u001b[39;49madd_basemap(ax, source\u001b[39m=\u001b[39;49mctxl\u001b[39m.\u001b[39;49mproviders\u001b[39m.\u001b[39;49mOpenStreetMap\u001b[39m.\u001b[39;49mMapnik)\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/loh/Workspace/methane_flux/notebook/retrieve_map.ipynb#X13sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m ax\u001b[39m.\u001b[39mtick_params(axis\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mboth\u001b[39m\u001b[39m\"\u001b[39m, which\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mmajor\u001b[39m\u001b[39m\"\u001b[39m, labelsize\u001b[39m=\u001b[39m\u001b[39m14\u001b[39m)\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/contextily/plotting.py:121\u001b[0m, in \u001b[0;36madd_basemap\u001b[0;34m(ax, zoom, source, interpolation, attribution, attribution_size, reset_extent, crs, resampling, **extra_imshow_args)\u001b[0m\n\u001b[1;32m    117\u001b[0m     left, right, bottom, top \u001b[39m=\u001b[39m _reproj_bb(\n\u001b[1;32m    118\u001b[0m         left, right, bottom, top, crs, {\u001b[39m\"\u001b[39m\u001b[39minit\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39m\"\u001b[39m\u001b[39mepsg:3857\u001b[39m\u001b[39m\"\u001b[39m}\n\u001b[1;32m    119\u001b[0m     )\n\u001b[1;32m    120\u001b[0m \u001b[39m# Download image\u001b[39;00m\n\u001b[0;32m--> 121\u001b[0m image, extent \u001b[39m=\u001b[39m bounds2img(\n\u001b[1;32m    122\u001b[0m     left, bottom, right, top, zoom\u001b[39m=\u001b[39;49mzoom, source\u001b[39m=\u001b[39;49msource, ll\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m\n\u001b[1;32m    123\u001b[0m )\n\u001b[1;32m    124\u001b[0m \u001b[39m# Warping\u001b[39;00m\n\u001b[1;32m    125\u001b[0m \u001b[39mif\u001b[39;00m crs \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/contextily/tile.py:222\u001b[0m, in \u001b[0;36mbounds2img\u001b[0;34m(w, s, e, n, zoom, source, ll, wait, max_retries)\u001b[0m\n\u001b[1;32m    220\u001b[0m x, y, z \u001b[39m=\u001b[39m t\u001b[39m.\u001b[39mx, t\u001b[39m.\u001b[39my, t\u001b[39m.\u001b[39mz\n\u001b[1;32m    221\u001b[0m tile_url \u001b[39m=\u001b[39m provider\u001b[39m.\u001b[39mbuild_url(x\u001b[39m=\u001b[39mx, y\u001b[39m=\u001b[39my, z\u001b[39m=\u001b[39mz)\n\u001b[0;32m--> 222\u001b[0m image \u001b[39m=\u001b[39m _fetch_tile(tile_url, wait, max_retries)\n\u001b[1;32m    223\u001b[0m tiles\u001b[39m.\u001b[39mappend(t)\n\u001b[1;32m    224\u001b[0m arrays\u001b[39m.\u001b[39mappend(image)\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/joblib/memory.py:594\u001b[0m, in \u001b[0;36mMemorizedFunc.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    593\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__call__\u001b[39m(\u001b[39mself\u001b[39m, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[0;32m--> 594\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_cached_call(args, kwargs)[\u001b[39m0\u001b[39m]\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/joblib/memory.py:537\u001b[0m, in \u001b[0;36mMemorizedFunc._cached_call\u001b[0;34m(self, args, kwargs, shelving)\u001b[0m\n\u001b[1;32m    534\u001b[0m         must_call \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    536\u001b[0m \u001b[39mif\u001b[39;00m must_call:\n\u001b[0;32m--> 537\u001b[0m     out, metadata \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcall(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    538\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmmap_mode \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    539\u001b[0m         \u001b[39m# Memmap the output at the first call to be consistent with\u001b[39;00m\n\u001b[1;32m    540\u001b[0m         \u001b[39m# later calls\u001b[39;00m\n\u001b[1;32m    541\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_verbose:\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/joblib/memory.py:779\u001b[0m, in \u001b[0;36mMemorizedFunc.call\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    777\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_verbose \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m    778\u001b[0m     \u001b[39mprint\u001b[39m(format_call(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfunc, args, kwargs))\n\u001b[0;32m--> 779\u001b[0m output \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfunc(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    780\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstore_backend\u001b[39m.\u001b[39mdump_item(\n\u001b[1;32m    781\u001b[0m     [func_id, args_id], output, verbose\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_verbose)\n\u001b[1;32m    783\u001b[0m duration \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mtime() \u001b[39m-\u001b[39m start_time\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/contextily/tile.py:252\u001b[0m, in \u001b[0;36m_fetch_tile\u001b[0;34m(tile_url, wait, max_retries)\u001b[0m\n\u001b[1;32m    250\u001b[0m \u001b[39m@memory\u001b[39m\u001b[39m.\u001b[39mcache\n\u001b[1;32m    251\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_fetch_tile\u001b[39m(tile_url, wait, max_retries):\n\u001b[0;32m--> 252\u001b[0m     request \u001b[39m=\u001b[39m _retryer(tile_url, wait, max_retries)\n\u001b[1;32m    253\u001b[0m     \u001b[39mwith\u001b[39;00m io\u001b[39m.\u001b[39mBytesIO(request\u001b[39m.\u001b[39mcontent) \u001b[39mas\u001b[39;00m image_stream:\n\u001b[1;32m    254\u001b[0m         image \u001b[39m=\u001b[39m Image\u001b[39m.\u001b[39mopen(image_stream)\u001b[39m.\u001b[39mconvert(\u001b[39m\"\u001b[39m\u001b[39mRGBA\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/contextily/tile.py:395\u001b[0m, in \u001b[0;36m_retryer\u001b[0;34m(tile_url, wait, max_retries)\u001b[0m\n\u001b[1;32m    375\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    376\u001b[0m \u001b[39mRetry a url many times in attempt to get a tile\u001b[39;00m\n\u001b[1;32m    377\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    392\u001b[0m \u001b[39mrequest object containing the web response.\u001b[39;00m\n\u001b[1;32m    393\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    394\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 395\u001b[0m     request \u001b[39m=\u001b[39m requests\u001b[39m.\u001b[39;49mget(tile_url, headers\u001b[39m=\u001b[39;49m{\u001b[39m\"\u001b[39;49m\u001b[39muser-agent\u001b[39;49m\u001b[39m\"\u001b[39;49m: USER_AGENT})\n\u001b[1;32m    396\u001b[0m     request\u001b[39m.\u001b[39mraise_for_status()\n\u001b[1;32m    397\u001b[0m \u001b[39mexcept\u001b[39;00m requests\u001b[39m.\u001b[39mHTTPError:\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget\u001b[39m(url, params\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m     \u001b[39mr\u001b[39m\u001b[39m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[39m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[39m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[39mreturn\u001b[39;00m request(\u001b[39m\"\u001b[39;49m\u001b[39mget\u001b[39;49m\u001b[39m\"\u001b[39;49m, url, params\u001b[39m=\u001b[39;49mparams, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[39m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[39m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[39m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[39mwith\u001b[39;00m sessions\u001b[39m.\u001b[39mSession() \u001b[39mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[39mreturn\u001b[39;00m session\u001b[39m.\u001b[39;49mrequest(method\u001b[39m=\u001b[39;49mmethod, url\u001b[39m=\u001b[39;49murl, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/requests/sessions.py:587\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    582\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    583\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m\"\u001b[39m: timeout,\n\u001b[1;32m    584\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    585\u001b[0m }\n\u001b[1;32m    586\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 587\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[1;32m    589\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/requests/sessions.py:701\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    698\u001b[0m start \u001b[39m=\u001b[39m preferred_clock()\n\u001b[1;32m    700\u001b[0m \u001b[39m# Send the request\u001b[39;00m\n\u001b[0;32m--> 701\u001b[0m r \u001b[39m=\u001b[39m adapter\u001b[39m.\u001b[39;49msend(request, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    703\u001b[0m \u001b[39m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    704\u001b[0m elapsed \u001b[39m=\u001b[39m preferred_clock() \u001b[39m-\u001b[39m start\n",
      "File \u001b[0;32m~/.local/conda/envs/gis/lib/python3.10/site-packages/requests/adapters.py:563\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    559\u001b[0m         \u001b[39mraise\u001b[39;00m ProxyError(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    561\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(e\u001b[39m.\u001b[39mreason, _SSLError):\n\u001b[1;32m    562\u001b[0m         \u001b[39m# This branch is for urllib3 v1.22 and later.\u001b[39;00m\n\u001b[0;32m--> 563\u001b[0m         \u001b[39mraise\u001b[39;00m SSLError(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    565\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    567\u001b[0m \u001b[39mexcept\u001b[39;00m ClosedPoolError \u001b[39mas\u001b[39;00m e:\n",
      "\u001b[0;31mSSLError\u001b[0m: HTTPSConnectionPool(host='a.tile.openstreetmap.org', port=443): Max retries exceeded with url: /19/262144/262143.png (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:997)')))"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsgAAAFsCAYAAADRzk8TAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAkhElEQVR4nO3dP2zUeX74/1d+OcVRJLxIKBLi1oSKSDuhYVPkprK0BegExWWlWGDpLingmoQiOco4YhsXG1Ocr8FuINKCXJBiWe3axUmuRtuAFHm9ElQWs2shRZZgXDnKab8Fsn/8eWED+4b3O8PjISEZZtb3+pyf6315mHnPH/3www8/BAAAEBER/1/tAQAAoCUWZAAAeIIFGQAAnmBBBgCAJ1iQAQDgCRZkAAB4ggUZAACesOeC3O/348KFCzEYDNLb5+fnY2FhIebn54sPBwAAb9tLLcjffvvtC2/r9/sxMTGx8zEAAPxf9pO97tDtdmNsbCy9bXFxMTqdTkREjI2NRa/Xi4mJiafuMxgMnnv0+X/+53+i3+/HkSNH4o//+I9fd3YAAHihP/zhD/Hf//3f8Vd/9Vfxp3/6py/9z+25IO/m4cOHO8vz6Oho3L9//7n7XLt2LX73u9/9mP8ZAAB4bZ999ln89V//9Uvf/0ctyPv3749Hjx7t/P7w4cPP3edXv/pV/OIXv3jqz77//vv45S9/GZ999lkcPHjwx4wAAACpBw8exOTkZPz5n//5K/1zP2pB7na78eWXX0ZExOrqapw7d+65+4yOjsbo6Gj6zx88eDDef//9HzMCQ2Zubi7Onz9fewwaogkyuiCjC17kVZ/Su+eL9BYXF+Obb76JK1euREREr9eLqampiIjodDqxf//+WFhYiE6n88LnKsPLOnXqVO0RaIwmyOiCjC4oZc9HkE+ePBknT57c+X23241ut7vz++xRY3hdIyMjtUegMZogowsyuqAUbxRCU65fv157BBqjCTK6IKMLSvmjH3744Ye3/T/63XffxUcffRS///3vPQcZAIA34nV3To8g05Tl5eXaI9AYTZDRBRldUIoFGQAAnmBBpinj4+O1R6AxmiCjCzK6oBQLMk2ZnZ2tPQKN0QQZXZDRBaVYkGnK2bNna49AYzRBRhdkdEEpFmSasrW1VXsEGqMJMrogowtKsSDTlC+++KL2CDRGE2R0QUYXlOIcZAAAhpJzkBkKS0tLtUegMZogowsyuqAUCzJN2bdvX+0RaIwmyOiCjC4oxYJMU7rdbu0RaIwmyOiCjC4oxYJMU2ZmZmqPQGM0QUYXZHRBKRZkmnL+/PnaI9AYTZDRBRldUIoFmaZsbGzUHoHGaIKMLsjoglIsyDRleXm59gg0RhNkdEFGF5TiHGQAAIaSc5AZCrdu3ao9Ao3RBBldkNEFpViQacqhQ4dqj0BjNEFGF2R0QSkWZJry4Ycf1h6BxmiCjC7I6IJSLMg0ZXp6uvYINEYTZHRBRheU4kV6NGVraytGRkZqj0FDNEFGF2R0wbO8SI+hsLa2VnsEGqMJMrogowtKsSDTlDt37tQegcZogowuyOiCUizINOXMmTO1R6AxmiCjCzK6oBQLMk25efNm7RFojCbI6IKMLijFgkxTjh49WnsEGqMJMrogowtKsSDTlGPHjtUegcZogowuyOiCUizINOXSpUu1R6AxmiCjCzK6oBTnIAMAMJScg8xQWFlZqT0CjdEEGV2Q0QWlWJBpyr1792qPQGM0QUYXZHRBKRZkmvLxxx/XHoHGaIKMLsjoglIsyDTlxo0btUegMZogowsyuqAUCzJNOX78eO0RaIwmyOiCjC4oxYJMU44cOVJ7BBqjCTK6IKMLSrEg05TLly/XHoHGaIKMLsjoglKcgwwAwFByDjJD4fbt27VHoDGaIKMLMrqgFAsyTVlfX689Ao3RBBldkNEFpViQacrp06drj0BjNEFGF2R0QSkWZJpy9erV2iPQGE2Q0QUZXVCKBZmmjI+P1x6BxmiCjC7I6IJSLMg05cCBA7VHoDGaIKMLMrqgFAsyTZmbm6s9Ao3RBBldkNEFpTgHGQCAoeQcZIZCr9erPQKN0QQZXZDRBaVYkGnK5uZm7RFojCbI6IKMLijFgkxTTpw4UXsEGqMJMrogowtKsSDTFC+w4FmaIKMLMrqgFAsyTTl16lTtEWiMJsjogowuKMWCTFNGRkZqj0BjNEFGF2R0QSkWZJpy/fr12iPQGE2Q0QUZXVCKc5ABABhKzkFmKCwvL9cegcZogowuyOiCUizIAADwhJ/sdYf5+fkYHR2NwWAQ586de+Ht9+/fj4sXL76RIXl3jI+P1x6BxmiCjC7I6IJSdn0Eud/vR7/fj4mJiZ2PX3T75uZmDAaDNzosw292drb2CDRGE2R0QUYXlLLrI8iLi4vR6XQiImJsbCx6vV5MTEzs3D42NhbffPNNzM/Px9jYWIyOjj73OQaDwXOL84MHD0rMzhA6e/Zs7RFojCbI6IKMLihl10eQHz58GO+9915ExM7TKJ41MTERKysrzz26vO3atWvx0UcfPfVrcnIyIiK+/vrriHj8E9/Gxkasr6/vvAvO0tJS9Hq9iIiYmZmJzc3NWFtbi6tXr0ZExK1bt+L27dsRETE9PR1bW1tx9+7duHHjRkRE3Lx5M1ZWViIi4tKlSxERsbKyEjdv3oyIiBs3bsTdu3dja2srpqenIyLi9u3bcevWrYiIuHr1aqytrcXm5mbMzMxERESv14ulpaWIePxuPevr67GxsbHzE+vy8vLOCwRc0+td09bW1tBd0zB+nd7mNW1tbQ3dNQ3j1+ltX9NXX301dNc0jF+nt31N33333dBd0zB+nd7mNX3++efxOnY95m37+cUTExOxsLAQEfHUI8i9Xi9WV1fj3LlzMTU1Fd1uN06ePPnU53jRI8iTk5OOeeM5c3Nzcf78+dpj0BBNkNEFGV3wrNc95m3Xp1h0u9348ssvIyJ2FuEnra6uxtjY2M59Hz169NznGB0dTZ96ARnf2HiWJsjogowuKGXXp1h0Op3Yv39/LCwsRKfT2Xke8tTUVET8/0+vWFhYiJWVlaceXYbXsf3XNLBNE2R0QUYXlLLnMW/PPmrc7Xaj2+1GxONHhx3tRkn79u2rPQKN0QQZXZDRBaV4oxCasv3DF2zTBBldkNEFpViQacr2q2JhmybI6IKMLijFgkxTvMCCZ2mCjC7I6IJSLMg0ZWNjo/YINEYTZHRBRheUYkGmKduHicM2TZDRBRldUMqubxTyprzuoc0AAPCyXnfn9AgyTdl+O0vYpgkyuiCjC0qxINOUQ4cO1R6BxmiCjC7I6IJSLMg05cMPP6w9Ao3RBBldkNEFpViQacr09HTtEWiMJsjogowuKMWL9GjK1tZWjIyM1B6DhmiCjC7I6IJneZEeQ2Ftba32CDRGE2R0QUYXlGJBpil37typPQKN0QQZXZDRBaVYkGnKmTNnao9AYzRBRhdkdEEpFmSacvPmzdoj0BhNkNEFGV1QigWZphw9erT2CDRGE2R0QUYXlGJBpinHjh2rPQKN0QQZXZDRBaVYkGnKpUuXao9AYzRBRhdkdEEpzkEGAGAoOQeZobCyslJ7BBqjCTK6IKMLSrEg05R79+7VHoHGaIKMLsjoglIsyDTl448/rj0CjdEEGV2Q0QWlWJBpyo0bN2qPQGM0QUYXZHRBKRZkmnL8+PHaI9AYTZDRBRldUIoFmaYcOXKk9gg0RhNkdEFGF5RiQaYply9frj0CjdEEGV2Q0QWlOAcZAICh5BxkhsLt27drj0BjNEFGF2R0QSkWZJqyvr5eewQaowkyuiCjC0qxINOU06dP1x6BxmiCjC7I6IJSLMg05erVq7VHoDGaIKMLMrqgFAsyTRkfH689Ao3RBBldkNEFpViQacqBAwdqj0BjNEFGF2R0QSkWZJoyNzdXewQaowkyuiCjC0pxDjIAAEPJOcgMhV6vV3sEGqMJMrogowtKsSDTlM3Nzdoj0BhNkNEFGV1QigWZppw4caL2CDRGE2R0QUYXlGJBpileYMGzNEFGF2R0QSkWZJpy6tSp2iPQGE2Q0QUZXVCKBZmmjIyM1B6BxmiCjC7I6IJSLMg05fr167VHoDGaIKMLMrqgFOcgAwAwlJyDzFBYXl6uPQKN0QQZXZDRBaVYkAEA4AkWZJoyPj5eewQaowkyuiCjC0qxINOU2dnZ2iPQGE2Q0QUZXVCKBZmmnD17tvYINEYTZHRBRheUYkGmKVtbW7VHoDGaIKMLMrqgFAsyTfniiy9qj0BjNEFGF2R0QSnOQQYAYCg5B5mhsLS0VHsEGqMJMrogowtKsSDTlH379tUegcZogowuyOiCUizINKXb7dYegcZogowuyOiCUizINGVmZqb2CDRGE2R0QUYXlGJBpinnz5+vPQKN0QQZXZDRBaVYkGnKxsZG7RFojCbI6IKMLihlzwV5fn4+FhYWYn5+/oX3WVxcjMXFxaKD8W5aXl6uPQKN0QQZXZDRBaXsuiD3+/3o9/sxMTGx8/GzthfnkydPvpkJeaf8/d//fe0RaIwmyOiCjC4oZdcFeXFxMTqdTkREjI2NRa/Xe+r2fr8fX3311c7HmcFgEN99991Tvx48eFBidobQrVu3ao9AYzRBRhdkdEEpuy7IDx8+jPfeey8iIkZHR+P+/ftP3d7r9eJnP/tZdLvdmJqaitXV1ec+x7Vr1+Kjjz566tfk5GRERHz99dcRETE7OxsbGxuxvr4ec3NzEfH4sO/thXxmZiY2NzdjbW0trl69GhGP/yW4fft2RERMT0/H1tZW3L17N27cuBERETdv3oyVlZWIiLh06VJERKysrMTNmzcjIuLGjRtx9+7d2Nraiunp6YiIuH379s6/XFevXo21tbXY3NzceVVsr9fbOYR8bm4u1tfXY2NjI2ZnZyPi8V/tbP/1jmt6vWs6dOjQ0F3TMH6d3uY1HTp0aOiuaRi/Tm/7mrYfaBmmaxrGr9PbvqY/+7M/G7prGsav09u8ps8//zxex65vNT0/Px+jo6MxMTERCwsLERExMTHxwtsHg0GcO3fuqc8xGAxiMBg89WcPHjyIyclJbzUNAMAb80bearrb7e48ary6uvrcAdxjY2M7tw8GgxgbG3vuc4yOjsb777//1K+DBw++9IC8W7Z/YoVtmiCjCzK6oJRdF+ROpxP79++PhYWF6HQ6O89DnpqaiojHL8zb3NyMhYWFePjwoRfq8aP98z//c+0RaIwmyOiCjC4o5Sd73eHZp0x0u92nHkn+5JNPyk/FO2ttbS3+8i//svYYNEQTZHRBRheU4o1CaMqdO3dqj0BjNEFGF2R0QSkWZJpy5syZ2iPQGE2Q0QUZXVCKBZmmbB/LAts0QUYXZHRBKRZkmnL06NHaI9AYTZDRBRldUIoFmaYcO3as9gg0RhNkdEFGF5RiQaYp2++CA9s0QUYXZHRBKbu+k96b8rrvagIAAC/rjbyTHrxt2++nDts0QUYXZHRBKRZkmnLv3r3aI9AYTZDRBRldUIoFmaZ8/PHHtUegMZogowsyuqAUCzJNuXHjRu0RaIwmyOiCjC4oxYJMU44fP157BBqjCTK6IKMLSrEg05QjR47UHoHGaIKMLsjoglIsyDTl8uXLtUegMZogowsyuqAU5yADADCUnIPMULh9+3btEWiMJsjogowuKMWCTFPW19drj0BjNEFGF2R0QSkWZJpy+vTp2iPQGE2Q0QUZXVCKBZmmXL16tfYINEYTZHRBRheUYkGmKePj47VHoDGaIKMLMrqgFAsyTTlw4EDtEWiMJsjogowuKMWCTFPm5uZqj0BjNEFGF2R0QSnOQQYAYCg5B5mh0Ov1ao9AYzRBRhdkdEEpFmSasrm5WXsEGqMJMrogowtKsSDTlBMnTtQegcZogowuyOiCUizINMULLHiWJsjogowuKMWCTFNOnTpVewQaowkyuiCjC0qxINOUkZGR2iPQGE2Q0QUZXVCKBZmmXL9+vfYINEYTZHRBRheU4hxkAACGknOQGQrLy8u1R6AxmiCjCzK6oBQLMgAAPMGCTFPGx8drj0BjNEFGF2R0QSkWZJoyOztbewQaowkyuiCjC0qxINOUs2fP1h6BxmiCjC7I6IJSLMg0ZWtrq/YINEYTZHRBRheUYkGmKV988UXtEWiMJsjogowuKMU5yAAADCXnIDMUlpaWao9AYzRBRhdkdEEpFmSasm/fvtoj0BhNkNEFGV1QigWZpnS73doj0BhNkNEFGV1QigWZpszMzNQegcZogowuyOiCUizINOX8+fO1R6AxmiCjCzK6oBQLMk3Z2NioPQKN0QQZXZDRBaVYkGnK8vJy7RFojCbI6IKMLijFOcgAAAwl5yAzFG7dulV7BBqjCTK6IKMLSrEg05RDhw7VHoHGaIKMLsjoglIsyDTlww8/rD0CjdEEGV2Q0QWlWJBpyvT0dO0RaIwmyOiCjC4oxYv0aMrW1laMjIzUHoOGaIKMLsjogmd5kR5DYW1trfYINEYTZHRBRheUYkGmKXfu3Kk9Ao3RBBldkNEFpViQacqZM2dqj0BjNEFGF2R0QSkWZJpy8+bN2iPQGE2Q0QUZXVCKBZmmHD16tPYINEYTZHRBRheUYkGmKceOHas9Ao3RBBldkNEFpey5IM/Pz8fCwkLMz8+/8D6DwSAuXLhQdDDeTZcuXao9Ao3RBBldkNEFpey6IPf7/ej3+zExMbHzcebKlSuxubn5Rgbk3fJv//ZvtUegMZogowsyuqCUXRfkxcXF6HQ6ERExNjYWvV4vvc/Pf/7zF36OwWAQ33333VO/Hjx48CPHZlitrKzUHoHGaIKMLsjoglJ2XZAfPnwY7733XkREjI6Oxv3795+6fTAYxKNHj2JsbOyFn+PatWvx0UcfPfVrcnIyIiK+/vrriIiYnZ2NjY2NWF9fj7m5uYiIWFpa2lnIZ2ZmYnNzM9bW1uLq1asREXHr1q24fft2RDx+a8mtra24e/du3LhxIyIev5J1+1+U7b9yWVlZ2XmF640bN+Lu3buxtbW189aUt2/fjlu3bkVExNWrV2NtbS02NzdjZmYmIiJ6vV4sLS1FRMTc3Fysr6/HxsZGzM7ORkTE8vJyLC8vu6YfcU337t0bumsaxq/T27yme/fuDd01DePX6W1f0+9///uhu6Zh/Dq97Wv6r//6r6G7pmH8Or3Na/r888/jdez6VtPz8/MxOjoaExMTsbCwEBERExMTO7d/+umnOx8vLS3Fb37zmzh58uRTn2MwGMRgMHjqzx48eBCTk5PeahoAgDfmjbzVdLfb3XnUeHV1Nbrd7lO3X7x4MS5evBi//vWvY2xs7LnlOOLxI8/vv//+U78OHjz40gPybtn+iRC2aYKMLsjoglJ2XZA7nU7s378/FhYWotPp7DwPeWpq6qn7LSwsxDfffPPCF/HByzp+/HjtEWiMJsjogowuKGXXp1i8Ka/7cDfDb2trK0ZGRmqPQUM0QUYXZHTBs97IUyzgbbt8+XLtEWiMJsjogowuKMUjyAAADCWPIDMUto9ygW2aIKMLMrqgFAsyTVlfX689Ao3RBBldkNEFpViQacrp06drj0BjNEFGF2R0QSkWZJqy/c45sE0TZHRBRheUYkGmKePj47VHoDGaIKMLMrqgFAsyTTlw4EDtEWiMJsjogowuKMWCTFPm5uZqj0BjNEFGF2R0QSnOQQYAYCg5B5mh0Ov1ao9AYzRBRhdkdEEpFmSasrm5WXsEGqMJMrogowtKsSDTlBMnTtQegcZogowuyOiCUizINMULLHiWJsjogowuKMWCTFNOnTpVewQaowkyuiCjC0qxINOUkZGR2iPQGE2Q0QUZXVCKBZmmXL9+vfYINEYTZHRBRheU4hxkAACGknOQGQrLy8u1R6AxmiCjCzK6oBQLMgAAPMGCTFPGx8drj0BjNEFGF2R0QSkWZJoyOztbewQaowkyuiCjC0qxINOUs2fP1h6BxmiCjC7I6IJSLMg0ZWtrq/YINEYTZHRBRheUYkGmKV988UXtEWiMJsjogowuKMU5yAAADCXnIDMUlpaWao9AYzRBRhdkdEEpFmSasm/fvtoj0BhNkNEFGV1QigWZpnS73doj0BhNkNEFGV1QigWZpszMzNQegcZogowuyOiCUizINOX8+fO1R6AxmiCjCzK6oBQLMk3Z2NioPQKN0QQZXZDRBaVYkGnK8vJy7RFojCbI6IKMLijFOcgAAAwl5yAzFG7dulV7BBqjCTK6IKMLSrEg05RDhw7VHoHGaIKMLsjoglIsyDTlww8/rD0CjdEEGV2Q0QWlWJBpyvT0dO0RaIwmyOiCjC4oxYv0aMrW1laMjIzUHoOGaIKMLsjogmd5kR5DYW1trfYINEYTZHRBRheUYkGmKXfu3Kk9Ao3RBBldkNEFpViQacqZM2dqj0BjNEFGF2R0QSkWZJpy8+bN2iPQGE2Q0QUZXVCKBZmmHD16tPYINEYTZHRBRheUYkGmKceOHas9Ao3RBBldkNEFpViQacqlS5dqj0BjNEFGF2R0QSnOQQYAYCg5B5mhsLKyUnsEGqMJMrogowtKsSDTlHv37tUegcZogowuyOiCUizINOXjjz+uPQKN0QQZXZDRBaVYkGnKjRs3ao9AYzRBRhdkdEEpFmSacvz48doj0BhNkNEFGV1QigWZphw5cqT2CDRGE2R0QUYXlGJBpimXL1+uPQKN0QQZXZDRBaU4BxkAgKHkHGSGwu3bt2uPQGM0QUYXZHRBKXsuyPPz87GwsBDz8/O73j41NRWDwaD4gLxb1tfXa49AYzRBRhdkdEEpuy7I/X4/+v1+TExM7Hz8pNXV1YiImJiYiIiIXq/3hsbkXXH69OnaI9AYTZDRBRldUMquC/Li4mJ0Op2IiBgbG3tuAe50OnHu3LmIeLxMb98XXtfVq1drj0BjNEFGF2R0QSk/2e3Ghw8fxtjYWEREjI6Oxv3799P7ra6uxgcffLBz3ycNBoPnnnrx4MGD152XITc+Pl57BBqjCTK6IKMLStn1EeT9+/fHo0ePdn5/+PDh5+7T7/fjyy+/jIsXL6af49q1a/HRRx899WtycjIiIr7++uuIiJidnY2NjY1YX1+Pubm5iIhYWlraecR6ZmYmNjc3Y21tbeenw1u3bu08GX96ejq2trbi7t27O++ic/PmzVhZWYmIiEuXLkVExMrKSty8eTMiHr/bzt27d2Nrayump6cj4vGT+2/duhURj38KXVtbi83NzZiZmYmIx08hWVpaioiIubm5WF9fj42NjZidnY2IiOXl5VheXnZNP+KaDhw4MHTXNIxfp7d5TQcOHBi6axrGr9Pbvqbt/34M0zUN49fpbV/T//7v/w7dNQ3j1+ltXtPnn38er2PXY95WV1d3lt+pqak4d+7cc48SX7hwIX7729/u/J/U7Xafuv1FjyBPTk465o3nzMzMxL/8y7/UHoOGaIKMLsjogme97jFve56DPD8/H6OjoxHx+MV4vV4vFhcX45NPPonFxcX493//9/jggw/iu+++i9/85jfPLcglhwUAgJf1ujvnrs9BjoidF+Ft63a7O0vwyZMn4+TJk684KrxY9rcQvNs0QUYXZHRBKd4ohKZsbm7WHoHGaIKMLsjoglIsyDTlxIkTtUegMZogowsyuqAUCzJN2X5VK2zTBBldkNEFpViQacqpU6dqj0BjNEFGF2R0QSkWZJoyMjJSewQaowkyuiCjC0qxINOU69ev1x6BxmiCjC7I6IJS9jwH+U1wDjIAAG/a6+6cHkGmKdtvRwnbNEFGF2R0QSkWZAAAeIIFmaaMj4/XHoHGaIKMLsjoglIsyDRldna29gg0RhNkdEFGF5RiQaYpZ8+erT0CjdEEGV2Q0QWlWJBpytbWVu0RaIwmyOiCjC4oxYJMU7744ovaI9AYTZDRBRldUIpzkAEAGErOQWYoLC0t1R6BxmiCjC7I6IJSLMg0Zd++fbVHoDGaIKMLMrqgFAsyTel2u7VHoDGaIKMLMrqgFAsyTZmZmak9Ao3RBBldkNEFpViQacr58+drj0BjNEFGF2R0QSkWZJqysbFRewQaowkyuiCjC0qxINOU5eXl2iPQGE2Q0QUZXVCKc5ABABhKzkFmKNy6dav2CDRGE2R0QUYXlGJBpimHDh2qPQKN0QQZXZDRBaVYkGnKhx9+WHsEGqMJMrogowtKsSDTlOnp6doj0BhNkNEFGV1Qihfp0ZStra0YGRmpPQYN0QQZXZDRBc/yIj2GwtraWu0RaIwmyOiCjC4oxYJMU+7cuVN7BBqjCTK6IKMLSrEg05QzZ87UHoHGaIKMLsjoglIsyDTl5s2btUegMZogowsyuqAUCzJNOXr0aO0RaIwmyOiCjC4oxYJMU44dO1Z7BBqjCTK6IKMLSrEg05RLly7VHoHGaIKMLsjoglKcgwwAwFByDjJDYWVlpfYINEYTZHRBRheUYkGmKffu3as9Ao3RBBldkNEFpViQacrHH39cewQaowkyuiCjC0qxINOUGzdu1B6BxmiCjC7I6IJSLMg05fjx47VHoDGaIKMLMrqgFAsyTTly5EjtEWiMJsjogowuKMWCTFMuX75cewQaowkyuiCjC0pxDjIAAEPJOcgMhdu3b9cegcZogowuyOiCUizINGV9fb32CDRGE2R0QUYXlGJBpimnT5+uPQKN0QQZXZDRBaVYkGnK1atXa49AYzRBRhdkdEEpFmSaMj4+XnsEGqMJMrogowtKsSDTlAMHDtQegcZogowuyOiCUizINGVubq72CDRGE2R0QUYXlOIcZAAAhpJzkBkKvV6v9gg0RhNkdEFGF5RiQaYpm5ubtUegMZogowsyuqAUCzJNOXHiRO0RaIwmyOiCjC4oxYJMU7zAgmdpgowuyOiCUizINOXUqVO1R6AxmiCjCzK6oBQLMk0ZGRmpPQKN0QQZXZDRBaXsuSDPz8/HwsJCzM/Pv9bt8CquX79eewQaowkyuiCjC0rZdUHu9/vR7/djYmJi5+NXuR1e1T/90z/VHoHGaIKMLsjoglJ+stuNi4uL0el0IiJibGwser1eTExMvPTtERGDwSAGg8FTf/b9999HRMSDBw9+/BUwVL7++uv4m7/5m9pj0BBNkNEFGV3wrO1d8w9/+MMr/XO7LsgPHz6MsbGxiIgYHR2N+/fvv9LtERHXrl2L3/3ud+nnn5ycfKVhAQDgVa2trcVf/MVfvPT9d12Q9+/fH48ePdr5/eHDh1/p9oiIX/3qV/GLX/ziqT+7f/9+/MM//EP8x3/8R/z0pz996WEZbg8ePIjJycn47LPP4uDBg7XHoQGaIKMLMrog8/3338cvf/nLnQd0X9auC3K3240vv/wyIiJWV1fj3Llzr3R7xONHlkdHR9PP/9Of/vSV3hebd8PBgwd1wVM0QUYXZHRB5k/+5E9e6f67vkiv0+nE/v37Y2FhITqdzs7zjKempl54OwAA/F+26yPIEZE+atztdl94OwAA/F/mjUIAAOAJVRbk0dHR+Md//McXPjeZd5MueJYmyOiCjC7IvG4Xf/TDDz/88IZmAgCA/3M8xQIAAJ5gQQYAgCfseYoFvAnz8/MxOjoag8EgPQll+/bV1dX4zW9+4zllAMBb88YfQZ6fn4+FhYWYn59/rdsZPv1+P/r9fkxMTOx8/KTV1dWIiJiYmIiIiF6v99ZnpI6X+X4wGAziwoULb3EqanuZLhYXF2NxcfEtTkVtL7tffPrpp295Mmrq9/tx4cKFGAwG6e0vu3e+0QV5r0Vor9sZTouLi9HpdCIidt585kmdTmfnUeV+v79zX4bby34/uHLlSmxubr7l6ajlZbrY/g/dyZMn3/Z4VPIq+8Xm5uYLlyWGT7/fj2+//faFt73s3vlGF+S9FqG9bmc4PXz4MN57772IeHz8yv3799P7ra6uxgcffOAdGt8RL/P9YHFxMX7+85+/7dGoaK8u+v1+fPXVVzsf827Yq4uxsbH45ptvYn5+PsbGxjxN7x3S7XZfuDe8yt75RhfkvRahl12UGC779++PR48e7fz+8OHDz92n3+/Hl19+GRcvXnybo1HRXt8PBoNBPHr0yA9M75i9uuj1evGzn/0sut1uTE1N7TxFi+H2MvvDxMRErKys+MGJHa+yd77RBXmvRehlFiWGT7fb3YlydXX1qbcu3/bpp5/uLMf+ZuHdsNf3gytXrsT9+/fjypUr0e/3Pd/0HbFXF4PBIA4fPhyjo6Nx8uRJ3y/eEXt10ev1YjAYxG9/+9uICN8viIhX2zvf6IK81yL0MosSw6fT6cT+/ftjYWEhOp3Ozl9zTE1NRcTjb2TffvttXLhwIf72b/+28rS8LXt9P7h48WJcvHgxfv3rX8fY2Jjnm74j9upibGxs5/bBYOBvGN4Re3Wxurq600K3231qKeLd9Sp75xt/J73t47oiHv91R6/Xi8XFxfjkk0/S24F3117fL7bvMzc3F//5n/9pGXpH7NXF1NRUdDqduH//vqdlvUN262IwGMSVK1fi8OHDunjHLC4uxr/+67/G3/3d38XFixdfe+/0VtMAAPAE76QHAABPsCADAMATLMgAAPAECzIAADzBggwAAE+wIAMAwBMsyAAA8AQLMgAAPOH/AVTmzojcfZtDAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 864x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot\n",
    "fig = plt.figure(figsize=(12, 6))\n",
    "ax = fig.add_subplot(111)\n",
    "\n",
    "extent = (-12600000, -10300000, 1800000, 3800000)\n",
    "ctxl.add_basemap(ax, source=ctxl.providers.OpenStreetMap.Mapnik)\n",
    "\n",
    "ax.tick_params(axis=\"both\", which=\"major\", labelsize=14)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 ('gis')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "8a10db6fd87fee191f68158fc4cc46992529de0a64044e82c33d618b82645974"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
